rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,833 @@
+implement.
+
+We also have `MaxCoalesceFusionRule`, which eagerly combines 'touching' candidates (ie candidates where the output of
+one sub-graph can be directly connected to the input of the other sub-graph)
+to form the largest possible candidate. The idea is once the search has been completed this rule can be used to collapse
+adjacent kernels intended for the same target.
+
+Here's some typical `FusionRule` combinations for different fusion strategies (please excuse the crudity of the diagram,
+I didn't have time to build it to scale or paint it):
+
+- Classic TVM `FuseOps`:
+
+```
+      OpCallByKindFusionRule
+                |
+                v
+    CombineByPrimitivesFusionRule (with default TVM primitive rules)
+```
+
+- Classic operator-based BYOC with `AnnotateTarget`/`MergeCompilerRegions`/`PartitionGraph` passes:
+
+```
+      OpPredicateFusionRule
+                |
+                v
+   CombineByPrimitivesFusionRule (with join anything primitive rule)
+```
+
+- Classic pattern-based BYOC with `MergeComposite`/`AnnotateTarget`/`PartitionGraph` passes:
+
+```
+     DFPatternFusionRule(pattern1)  ...  DFPatternFusionRule(patternn)
+                |                                    |
+                v                                    v
+      CompositeFusionRule(label1)   ...   CompositeFusionRule(labeln)
+                        \                     /
+                         v                   v
+                            UnionFusionRule
+                                   |
+                                   v
+                     CombineByPrimitivesFusionRule (with join anything primitive rule)
+```
+
+- "Just fuse what I tell you to fuse", using `DFPatterns` to directly select candidates:
+
+```
+     DFPatternFusionRule(pattern1)  ...  DFPatternFusionRule(patternn)
+                           \                 /
+                            v               v
+                             UnionFusionRule
+```
+
+- "Consider this library implementation for these sub-expressions", using `DFPatterns` to pick out which Relay operators
+  are supported (note that TVM lowering does not currently support this):
+
+```
+    OpCallByKindFusionRule     DFPatternFusionRule(pattern1) ... DFPatternFusionRule(patternn)
+                    \                       |                                 |
+                     \                      v                                 v
+                      \         CompositeFusionRule(label1)  ...  CompositeFusionRule(labeln)
+                       \                    |                        /
+                        v                   v                       v
+                                      UnionFusionRule
+                                            |
+                                            v
+                             CombineByPrimitivesFusionRule (with default TVM primitive rules)
+```
+
+### FusionSpec
```

Comment:
I'm warming up to your suggestion (thank you!) to characterize this as 'partitioning' rather than 'fusion', in which I'd rename throughout (FusionSpec -> PartitionSpec, CandidateKernel -> CandidatePartition, CollageFuseOps -> CollagePartition etc). As part of that I could move CollagePartition earlier in the standard pass prefix, and be prepared to run target-specific passes after partitioning but before lowering. This is all more TVM surgery than I was anticipating but I'll give it a go in the prototype and report back.